packages/block-editor/src/hooks/supports.js

```diff
@@ -339,3 +339,33 @@ export const getLayoutSupport = ( nameOrType ) =>
  */
 export const hasStyleSupport = ( nameOrType ) =>
 	styleSupportKeys.some( ( key ) => hasBlockSupport( nameOrType, key ) );
+
+/**
+ * Returns true if the block defines support for block class name.
+ *
+ * @param {string|Object} nameOrType Block name or type object.
+ * @return {boolean} Whether the block supports the feature.
+ */
+export const hasBlockClassNameSupport = ( nameOrType ) => {
+	const classNameSupport = getBlockSupport( nameOrType, 'className', true );
+
+	if ( classNameSupport === true || classNameSupport === false ) {
+		return classNameSupport;
+	}
+
+	// classNameSupport can be an object. If it doesn't have a block key,
+	// we default to true.
+	return (
+		! Object.hasOwn( classNameSupport, 'block' ) ||
+		classNameSupport.block === true
+	);
+};
+
+/**
+ * Returns true if the block defines support for variation class name.
+ *
+ * @param {string|Object} nameOrType Block name or type object.
+ * @return {boolean} Whether the block supports the feature.
+ */
+export const hasVariationClassNameSupport = ( nameOrType ) =>
```

Comment:
Should we default to `true` for this feature? Have you discussed this? This would also affect the default handling when we have `supports.className: true`.

Reply:
I was originally hoping to do that (and discussed it with @gziolo again today), but I've come to the conclusion that it's too much of a risk. As you mentioned [elsewhere](https://github.com/WordPress/gutenberg/pull/61864#issuecomment-2236340867), setting this to `true` will change the serialized markup for existing blocks, so we'll need to add migrations for them.

Having to add those for any block with variations that wants to use this new feature seemed like a bit of a hassle at first, but I now think it's an okay tradeoff. It's probably worse for plugins (and authors) that include a large number of blocks -- and for Core itself -- but even then, not all blocks come with variations, and it's a one time change. IMO, it's reasonable to ask extenders to enable this feature manually (and to add a deprecation) in exchange for a useful enhancement.

Since we might also need [a server-side part](https://github.com/WordPress/gutenberg/pull/61864#issuecomment-2233381017) (which will require a robust way of [determining the active variation](https://github.com/WordPress/wordpress-develop/pull/6602)), it makes even more sense to require extenders to opt into this new feature explicitly, as it gives them control over adding this new class name, and allows them to fix their `isActive` fields before they do so.